### DISCUSSION TITLE
Enhancing icalendar through attributes

### DISCUSSION THREAD

**Discussion by niccokunzmann** [collective/icalendar#662] [Ideas]
The icalendar standard has clarity about which attributes can be expected in components.

Attribtues can
- (S1) occur mandatory
- (S2) occur optionally

They can also
- (S3) occur at most once
- (S4) occur several times

Several issues exist that hint at a better API that we can provide.

## (P1) Proposal for (S1) mandatory properties

The `UID` attribute of VEVENT is an example of this.

It is mandatory and also occurs once.

Add mandatory attributes to the constructor as well as a `make_valid=True` argument.
In case that argument is set, mandatory attributes receive a valid value.

```python
Event() # sets UID, DTSTART to now.
Event(make_valid=False) # sets nothing
Event(uid="{aaaa-aaaa-aaaa-aaaa}") # sets the UID to the value and DTSTART, too.
```

We need `make_valid=False` in order to create the components `from_ical()`

## Proposal for S1 & S2 - accessing properties

Attributes that are defined in the standard can be added as properties in Pythonic form.
The setter allows setting values that are supported as argument to the wrapping icalendar property.

```python

class MandatoryPropertyNotSet(ValueError):
    """A mandatory property was not set on a component."""


class Event:

    @property
    def dtstart(self) -> vDatetime:  # all supported types
        """The DTSTART property (mandatory)

        See `RFC 5545 ...<link to rfc>`

        If DTSTART is not set, this raises and MandatoryPropertyNotSet error.
        """

    @dtstart.setter
    def dtstart(self, value: Union[vDatetime, datetime, date]):
        # ...
```

## (P2) Proposal for S3 & S1 - properties that occur once

When attributes can occur only once, we can set them and access them in that way.
See above for the `dtstart` example.

## (P3) Proposal for S3 & S2 - properties that occur several times

An attribute that can occur once at most and is optional, should look like this:

```python
@property
def summary(self) -> Optional[vText]:
    """The SUMMARY (optional)

    See `RFC 5545 ...<link to rfc>`
    """

@dtstart.setter
def summary(self, value: Union[vText, str, bytes, None]):
    # ...
```

## (P3) Proposal for S4 - attributes occur more than once

Issues: #224
It creates edge cases that are rarely caught if we allow attributes to be

- a list like vDDDLists
- a value like vDDDTypes at the same time.

Proposal to access those attributes all as a list:

```python
@property
def rrules(self) -> vList[vRRule]: # plural
    """The RRULE

    See `RFC 5545 ...<link to rfc>`
    """

@rrules.setter
def rrules(self, value: vList[vRRule]]):
    # ...
```

Here, it can be confusing if one beliefs that this can be modified.
The project [ObservableList](https://observablelist.readthedocs.io/en/latest/reference/ObservableList/init.html) handles this use-case by creating a list that
can be observed for changes. These can be then copied into the component.

```python
event = Event()
event.rrules.append(str | bytes | vRRule | dateutil.rrule.rrule)
```
## Requirements

I would like the following to happen:

- properties are types
- properties are set+get
- properties validate the input and convert it to a useful one
- properties refer to the RFC

## Related

- #596 
- #315
- #431

## Discussion

I wonder what you think about this, especially:

- What would be really good behavior?
- What would you expect to be in there?
- What edge cases come up for you?

**Comment by SemTiOne** [comment]
For #224, what if we normalize `__getitem__` in `component.py` line 231 so `event["EXDATE"]` and `event["RDATE"]` always return a list even with a single line?

**Reply by niccokunzmann** [reply]
I decided to keep the dictionairy interface and not add special cases.
The next thing to think about then is mutability of the content, pop() and other dict functions that need to be tested, also in combination.
Should then setitem be changed instead?

Would backward compatibility to version 4 break with that? Icalendar has been around for years. While we make it nice now, preserving old software and it being able to work with the latest version is of value. I opted for that and was hesitant to touch the core code. Ow, there sre more people working on it, so it might be time to re-evaluate the strategy.

Should we generalize it? If generalized, what is with RRULE, usually only appearing once. Components have a registry of what appears multiple times.

Let me know what you think and if you would stick with and through evaluating and driving such a change.

**Reply by niccokunzmann** [reply]
We also progressed a bit further. There is documentation on the design decisions.

**Reply by stevepiercy** [reply]
> We also progressed a bit further. There is documentation on the design decisions.

See:

- https://icalendar.readthedocs.io/en/latest/explanation/api-design.html
- https://github.com/collective/icalendar/issues/1662

**Reply by SemTiOne** [reply]
> I decided to keep the dictionairy interface and not add special cases. The next thing to think about then is mutability of the content, pop() and other dict functions that need to be tested, also in combination. Should then setitem be changed instead?
> 
> Would backward compatibility to version 4 break with that? Icalendar has been around for years. While we make it nice now, preserving old software and it being able to work with the latest version is of value. I opted for that and was hesitant to touch the core code. Ow, there sre more people working on it, so it might be time to re-evaluate the strategy.
> 
> Should we generalize it? If generalized, what is with RRULE, usually only appearing once. Components have a registry of what appears multiple times.
> 
> Let me know what you think and if you would stick with and through evaluating and driving such a change.

Leave `__getitem__`/`add()` untouched. The dict interface has been stable since v4.0 (api-design.rst), so changing it goes through the `deprecate_for_version_8` path. I believe it's not worth it for this.

On mutability, the proposed fix helps the read side, but the write side stays messy. With 3+ occurrences, `event['EXDATE']` returns the same list stored internally. `.pop()/.append()` mutates the component without validation:
```python
e = Event()
e.add('EXDATE', datetime(2024,1,1))
e.add('EXDATE', datetime(2024,1,2))
e.add('EXDATE', datetime(2024,1,3))
val = e['EXDATE']
val.pop()
val is e['EXDATE'] # True, same object
len(e['EXDATE']) # 2, mutated in place 
```
With 1 occurrence, the fix returns a new list each call. `.pop()` works but mutates a disconnected copy, not the stored value. Either way, no guardrails at the dict level. The computed properties do: `event.exdates` returns a disconnected snapshot (per its docstring: "Modifying the returned list does not change the EXDATE value"), so the footgun doesn't exist there.

Btw, I re-checked and my original ask already solves a layer up, for the read side at least. rdates/exdates/rrules normalize to a plain list unconditionally in attr.py, and rdates/exdates are writable now too. Same pattern already wires on the base `Component` class for `comments`, `links`, `related_to`, and `concepts` and on Event/Todo specifically for `categories`, `attendees`, and `contacts`. So `event["EXDATE"]` staying inconsistent at the dict level is fine as long as `event.exdates` is the documented, safe way in.

On generalizing, I'd hold off. The property access layer handles it fine already. The registry itself has at least one stale entry. `Event.multiple` lists `"RELATED"`, but the RFC 5545 property (and the `related_to` getter) is keyed on `RELATED-TO`, so that entry never matches the parsed data. Auto-generating properties off a registry with drift like that, on top of the mutability quirks, seems risky right now.

After checking every multiple entry on Event/Todo against what's wired (including inherited base-Component properties), the remaining gap is just ATTACH, RSTATUS, and RESOURCES. Everything else on those two components is already covered, following the exact non-live-list getter/setter/deleter pattern already in attr.py (so the mutability footgun above doesn't leak into the new properties either). No changes to `add()`/`__setitem__`/the `multiple` registry itself. 

Yes I'll stick with it. What do you think? @niccokunzmann

**Reply by niccokunzmann** [reply]
Did you AI generate the answer?

**Reply by SemTiOne** [reply]
For writing the answer, no. For research and connect things, yes I use AI as a discussion partner.

**Reply by niccokunzmann** [reply]
Thank you! That is important. We get so many AI contributions these days and I want to be sure that if you intend to touch the core logic, that you trust yourself to do so and not the AI. AI is really bad at quality and understanding intention at present.

**Reply by niccokunzmann** [reply]
Uh. RELATED is probably be a spelling mistake if it does not occur in the RFC.

RSTATUS and RESOURCES - if you like, open the issue. We should cover all standardized attributes.

For attach, we have a PR open at the moment (or maybe merged?), adding the ATTACH and attachments properties.

Thanks for looking for completeness.

I would still like the resulting list to be mutable and write back. That would be a huge improvement, I would say. However, I am not sure how to achieve this in some cases. There seem to be always edge cases that break when I think about implementation.

**Reply by stevepiercy** [reply]
- [`RELATED`](https://datatracker.ietf.org/doc/html/rfc5545#section-3.2.14) is a parameter for an `ALARM` trigger.
- [`RELATED-TO`](https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.4.5) is a property for either `VEVENT`, `VTODO`, or `VJOURNAL` calendar components.

The RFC's are tricky.
